feat(cloud-sql): add postgres connector sample (#3266)
GoogleCloudPlatform/nodejs-docs-samples: cloud-sql/postgres/knex/README.md

[Google Cloud console](https://console.cloud.google.com/sql) or by running
the following command:

```sh
gcloud sql instances describe <INSTANCE_NAME> --format='value(connectionName)'
```

#### Linux / Mac OS

Use these terminal commands to initialize environment variables:

```bash
export INSTANCE_CONNECTION_NAME='<INSTANCE_CONNECTION_NAME>'
export DB_USER='<DB_USER_NAME>'
export DB_PASS='<DB_PASSWORD>'
export DB_NAME='<DB_NAME>'
```

#### Windows/PowerShell

Use these PowerShell commands to initialize environment variables:

```powershell
$env:INSTANCE_CONNECTION_NAME="<INSTANCE_CONNECTION_NAME>"
$env:DB_USER="<DB_USER_NAME>"
$env:DB_PASS="<DB_PASSWORD>"
$env:DB_NAME="<DB_NAME>"
```

### Testing the application

1. Next, install the Node.js packages necessary to run the app locally by
   running the following command:

    ```sh
    npm install
    ```

2. Run the sample app locally with the following command:

    ```sh
    npm start
    ```

Navigate towards `http://127.0.0.1:8080` to verify your application is running
correctly.

## Deploy to Google App Engine Standard

1. To allow your app to connect to your Cloud SQL instance when the app is

>    deployed, add the user, password, database, and instance unix socket

Reviewer: ```suggestion
   deployed, add the user, password, database, and instance connection name
```

   variables from Cloud SQL to the related environment variables in the
   [`app.standard.yaml`](app.standard.yaml) file.

    ```yaml
    env_variables:
      INSTANCE_CONNECTION_NAME: <MY-PROJECT>:<INSTANCE-REGION>:<INSTANCE-NAME>
      DB_USER: MY_DB_USER
      DB_PASS: MY_DB_PASSWORD
      DB_NAME: MY_DATABASE
    ```

2. To deploy to App Engine Standard, run the following command:

    ```sh
    gcloud app deploy app.standard.yaml
    ```

3. To launch your browser and view the app at
   <https://[YOUR_PROJECT_ID]>.appspot.com, run the following command:

    ```sh
    gcloud app browse
    ```

## Deploy to Google App Engine Flexible

> 1. Add the user, password, database, and instance unix socket variables from

Reviewer: ```suggestion
1. Add the user, password, database, and instance connection name variables from
```

Cloud SQL to the related environment variables in the
[`app.flexible.yaml`](app.flexible.yaml) file.

    ```yaml
    env_variables:
      INSTANCE_CONNECTION_NAME: <MY-PROJECT>:<INSTANCE-REGION>:<INSTANCE-NAME>
      DB_USER: MY_DB_USER
      DB_PASS: MY_DB_PASSWORD
      DB_NAME: MY_DATABASE
    ```

2. To deploy to App Engine Node.js Flexible Environment, run the following
   command:

    ```sh
    gcloud app deploy app.flexible.yaml
    ```

3. To launch your browser and view the app at
   <https://[YOUR_PROJECT_ID]>.appspot.com, run the following command:

    ```sh
    gcloud app browse
    ```

## Deploy to Cloud Run

See the [Cloud Run
documentation](https://cloud.google.com/sql/docs/postgres/connect-run) for more
details on connecting a Cloud Run service to Cloud SQL.

Build and deploy the service to Cloud Run:

```sh
gcloud run deploy run-sql --source . /
    --allow-unauthenticated \
    --set-env-vars INSTANCE_CONNECTION_NAME=[INSTANCE_CONNECTION_NAME],\
      DB_USER=[MY_DB_USER],DB_PASS=[MY_DB_PASS],DB_NAME=[MY_DB]
```

Replace environment variables with the correct values for your Cloud SQL
instance configuration.

> This step can be done as part of deployment but is separated for clarity.

Reviewer: ```suggestion
```

      DB_USER: MY_DB_USER
      DB_PASS: MY_DB_PASSWORD
      DB_NAME: MY_DATABASE
    ```

2. To deploy to App Engine Node.js Flexible Environment, run the following
   command:

    ```sh
    gcloud app deploy app.flexible.yaml
    ```

3. To launch your browser and view the app at
   <https://[YOUR_PROJECT_ID]>.appspot.com, run the following command:

    ```sh
    gcloud app browse
    ```

## Deploy to Cloud Run

See the [Cloud Run
documentation](https://cloud.google.com/sql/docs/postgres/connect-run) for more
details on connecting a Cloud Run service to Cloud SQL.

1. Build the container image:

```sh
gcloud builds submit --tag gcr.io/[YOUR_PROJECT_ID]/run-sql
```

1. Deploy the service to Cloud Run:

```sh
gcloud run deploy run-sql --image gcr.io/[YOUR_PROJECT_ID]/run-sql
```

Take note of the URL output at the end of the deployment process.

1. Configure the service for use with Cloud Run

```sh
gcloud run services update run-sql \
    --add-cloudsql-instances [INSTANCE_CONNECTION_NAME] \
    --set-env-vars INSTANCE_CONNECTION_NAME=[INSTANCE_CONNECTION_NAME],\
      DB_USER=[MY_DB_USER],DB_PASS=[MY_DB_PASS],DB_NAME=[MY_DB]

> ```

Author: I don't know `gcloud` well enough to have an opinion but having some experience with docker containers in the past, it seems quite reasonable to me to have distinct deploy / run commands. Even if the intention in this case is just to raise API / commands awareness.

Replace environment variables with the correct values for your Cloud SQL
instance configuration.

This step can be done as part of deployment but is separated for clarity.

It is recommended to use the [Secret Manager
integration](https://cloud.google.com/run/docs/configuring/secrets) for Cloud
Run instead of using environment variables for the SQL configuration. The
service injects the SQL credentials from Secret Manager at runtime via an
environment variable.

Create secrets via the command line:

```sh

> echo -n $INSTANCE_UNIX_SOCKET | \
>     gcloud secrets create [INSTANCE_UNIX_SOCKET_SECRET] --data-file=-
> ```

Reviewer: should be `INSTANCE_CONNECTION_NAM` for Connector usage
```suggestion
echo -n $INSTANCE_CONNECTION_NAME | \
    gcloud secrets create [INSTANCE_CONNECTION_NAME_SECRET] --data-file=-
```
```

Deploy the service to Cloud Run specifying the env var name and secret name:

```sh

> gcloud beta run deploy SERVICE --image gcr.io/[YOUR_PROJECT_ID]/run-sql \
>     --add-cloudsql-instances $INSTANCE_CONNECTION_NAME \
>     --update-secrets DB_USER=[DB_USER_SECRET]:latest, \
>       DB_PASS=[DB_PASS_SECRET]:latest, \
>       DB_NAME=[DB_NAME_SECRET]:latest

Reviewer: update command
```suggestion
gcloud run deploy SERVICE --image gcr.io/[YOUR_PROJECT_ID]/run-sql \
    --update-secrets INSTANCE_CONNECTION_NAME=[INSTANCE_CONNECTION_NAME_SECRET]:latest, \
      DB_USER=[DB_USER_SECRET]:latest, \
      DB_PASS=[DB_PASS_SECRET]:latest, \
      DB_NAME=[DB_NAME_SECRET]:latest
```

```

1. Navigate your browser to the URL noted in step 2.

For more details about using Cloud Run see <http://cloud.run>. Review other
[Node.js on Cloud Run samples](../../../run/).

## Deploy to Cloud Functions

To deploy the service to [Cloud Functions](https://cloud.google.com/functions/docs) run the following command:

```sh
gcloud functions deploy votes --gen2 --runtime nodejs18 --trigger-http \
  --allow-unauthenticated \
  --entry-point votes \
  --region <INSTANCE_REGION> \
  --set-env-vars INSTANCE_CONNECTION_NAME=<PROJECT_ID>:<INSTANCE_REGION>:<INSTANCE_NAME> \
  --set-env-vars DB_USER=$DB_USER \
  --set-env-vars DB_PASS=$DB_PASS \
  --set-env-vars DB_NAME=$DB_NAME
```

> Note: If the function fails to deploy or returns a `500: Internal service error`,
> this may be due to a known limitation with Cloud Functions gen2 not being able
> to configure the underlying Cloud Run service with a Cloud SQL connection.
> 
> A workaround command to fix this is is to manually revise the Cloud Run
> service with the Cloud SQL Connection:
> 
> ```sh
> gcloud run deploy votes --source . \
>   --region <INSTANCE_REGION> \
>   --add-cloudsql-instances <PROJECT_ID>:<INSTANCE_REGION>:<INSTANCE_NAME>
> ```
> 
> The Cloud Function command above can now be re-run with a successful deployment.

Reviewer: this issue is with proxy only, not with Connector (why recommending Connector usage is important)
```suggestion
```
